lib/auth/grpcserver.go

```diff
@@ -1915,22 +1925,59 @@ func (g *GRPCServer) DeleteAllKubernetesServers(ctx context.Context, req *proto.
 	return &emptypb.Empty{}, nil
 }
 
+var MinSupportedRoleV7Version = semver.New(utils.VersionBeforeAlpha("14.0.0"))
+
+// maybeDowngradeRole tests the client version passed through the GRPC metadata, and
+// if the client version is unknown or less than the minimum supported version
+// for V7 roles returns a shallow copy of the given role downgraded to V6, If
+// the passed in role is already V6, it is returned unmodified.
+func maybeDowngradeRole(ctx context.Context, role *types.RoleV6) (*types.RoleV6, error) {
+	if role.Version != types.V7 {
+		// role is already <V7, no need to downgrade
+		return role, nil
+	}
+
+	var clientVersion *semver.Version
+	clientVersionString, ok := metadata.ClientVersionFromContext(ctx)
+	if ok {
+		var err error
+		clientVersion, err = semver.NewVersion(clientVersionString)
+		if err != nil {
+			return nil, trace.BadParameter("unrecognized client version: %s is not a valid semver", clientVersionString)
+		}
+	}
+
+	if clientVersion == nil || clientVersion.LessThan(*MinSupportedRoleV7Version) {
+		log.Debugf(`Client version "%s" is unknown or less than 14.0.0, converting role to v6`, clientVersionString)
```

Comment:
I don't think we should downgrade if the client version is unknown. Teleport clients have been reporting their versions for long enough now that if they don't, they probably don't support v6 roles anyway. This is more likely to be a third-party client and downgrading the role in that case will just be confusing I think